add special types for racket/private/template

bennn opened the pull request:
Add type definitions for the new 'Guide' data structure.

@rmculpepper can you double-check these? 

Closes #688

samth commented:
Thanks for getting to this!

bennn commented:
@stamourv I think this test failure means there's something wrong with the optimizer's dead-code elimination.

```
The command "if [ $TESTS = "int"  ]; then racket -l typed-racket-test -- --int; fi" exited with 0.
334.30s$ if [ $TESTS = "int"  ]; then racket -l typed-racket-test -- --opt; fi
--------------------
Log Comparison for dead-case-lambda.rkt >  > Optimization Tests > Typed Racket Tests > Unnamed test
FAILURE
name:        check-equal?
location:    typed-racket-test/optimizer/run.rkt:28:6
expression:  (check-equal? log expected-log)
params:
  '(("TR opt: dead-case-lambda.rkt 3:7 (case-lambda (() (void)) ((d) (void)) ((d . rst) (void))) -- dead case-lambda branch"
     "TR opt: dead-case-lambda.rkt 6:10 (d . rst) -- dead case-lambda branch")
    ("TR opt: dead-case-lambda.rkt 4:10 () -- dead case-lambda branch"
     "TR opt: dead-case-lambda.rkt 6:10 (d . rst) -- dead case-lambda branch"))
actual:
  '("TR opt: dead-case-lambda.rkt 3:7 (case-lambda (() (void)) ((d) (void)) ((d . rst) (void))) -- dead case-lambda branch"
    "TR opt: dead-case-lambda.rkt 6:10 (d . rst) -- dead case-lambda branch")
expected:
  '("TR opt: dead-case-lambda.rkt 4:10 () -- dead case-lambda branch"
    "TR opt: dead-case-lambda.rkt 6:10 (d . rst) -- dead case-lambda branch")
--------------------
165 success(es) 1 failure(s) 0 error(s) 166 test(s) run
Typed Racket Tests did not pass.
```

Here's the function (from `optimizer/tests/dead-case-lambda.rkt`); it looks like the first case isn't getting added to the `lambda-dead-table`, and this seems to happen on the `master` branch too.

```
#lang typed/racket
(procedure-arity
  (ann (case-lambda
         [() (void)]
         [(d) (void)]
         [(d . rst) (void)])
          (Any -> Any)))
```

rmculpepper commented:
`t-const` and `t-var` are actually bound to `#%expression`, so they shouldn't need types. Similarly, `t-list` and `t-list*` are bound to `list` and `list*`.

`t-append` is essentially `append` except with a shortcut if the second list is empty, but beware that it allows improper lists. `t-orelse` is a macro; it's run-time helper `t-orelse*` takes two thunks and returns the result of one of them. I think in practice the return type is either `Syntax` or `(Listof Syntax)`.

Here are the types of functions that I think occur in expanded code:

    Stxish = '() | Syntax | (cons Syntax Stxish)
    t-append : (Listof Syntax) Stxish -> Stxish
    t-resyntax : (U Syntax #f) Syntax Any -> Syntax
    t-relocate : Syntax (U Syntax #f) -> Syntax
    t-orelse* : forall a, (-> a) (-> a) -> a
    t-struct : Any (Listof Syntax) -> Syntax
    h-splice : Any (U Syntax #f) (U Syntax #f) -> (Listof Syntax)
    t-subst : (U Syntax #f) Syntax X Y ... -> Syntax
      where X, Y are somewhat complicated
    check-same-length : (U Syntax #f) (U Syntax #f) (Listof Any) -> Void

AlexKnauth commented:
What does `make-template-identifier` do?

samth commented:
It solves the problem with non-provided identifiers that your other issue
mentioned.

On Sun, Apr 15, 2018, 12:07 AM Alex Knauth <notifications@github.com> wrote:

> What does make-template-identifier do?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/racket/typed-racket/pull/690#issuecomment-381378658>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO785qK6M3utA73YQpNBBxhcuRQP5-aks5tosdtgaJpZM4TL3yZ>
> .
>

samth commented:
Can we merge this soon? What do we still need?

stamourv commented:
@bennn: Re dead case-lambda branch: Looking at the test results, I think it's just a different source location; the test itself is still morally passing. So it should be ok to just fix the test to expect the new output, I think.

bennn commented:
Updated with Ryan's types and changed the failing `case-lambda` test.

(Vincent after looking again, I agree the test is morally passing. Probably we should file an issue for about the changed source location.)
